### DISCUSSION TITLE
Slimming down yamcs docker images

### DISCUSSION THREAD

**Discussion by jbell-astrobotic** [yamcs/yamcs#842] [General]
Hi! I'd like to discuss ways to slim down docker images built with maven. Generally, with build infrastructure (like maven) a lot of stuff is generated that is not actually needed to run the project - this results in bloated images that hinder download, upload, and build speeds and also increases the attack surface of an image. Ideally, I'd like to set up my Dockerfile closer to best practices where it looks something like:

FROM maven and jdk image as build
--- do build ---
FROM jre 11 image
COPY --from build <jar file>
CMD ["java" "-jar" "something.jar"]

After a few attempts to do this, I've not had success. Was looking to start discussions on how to get away from maven for running the project and simply use it for building a fat jar. 

I think I'm just missing all of the additional steps that yamcs:run is doing, and that's where I'm blocked at right now. I've tried re-configuring the pom to create a fat jar and remove the .tar.gz zipping in packaging and sort of guessed that the main class should be org.yamcs.YamcsServer, but yeah getting linking issues and various errors.

Would love to discuss how to get this kind of functionality working, and if devs need collaboration let me know - happy to work with you on this.

**Comment by merose** [comment]
Why don't you just do `mvn package` and then take the tar-ball from the target directory. You can install that in a fresh image. It only contains the distribution, no source, no .git/. Or you can customize by writing a descriptor for the Maven assembly plugin if you don't like the default packaging from the Yamcs plugin.

It's not sufficiently streamlined to use with `java -jar`, but Yamcs also requires additional configuration which is less amenable to `-jar` anyway.

**Reply by jbell-astrobotic** [reply]
This would still require maven and the maven jdk image as the base layer, no? So in regards to slimming down the production image, the maven:3.6-openjdk-11 image is 658MB, whereas something like the eclipse-temurine:11.0.21_9-jre-alpine image which is an example of a minimal image to run a jar file is 155MB - over 4 times smaller.

**Comment by merose** [comment]
No, I'm talking about building either outside of an image, or creating an
image just for building, then extracting the .tar.gz file and applying that
to a fresh image. In neither case would you need anything but Java in the
destination image.

Mark

On Wed, Nov 1, 2023 at 10:02 AM jbell-astrobotic ***@***.***>
wrote:

> This would still require maven and the maven jdk image as the base layer,
> no? So in regards to slimming down the production image, the
> maven:3.6-openjdk-11 image is 658MB, whereas something like the
> eclipse-temurine:11.0.21_9-jre-alpine image which is an example of a
> minimal image to run a jar file is 155MB - over 4 times smaller.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/yamcs/yamcs/discussions/842#discussioncomment-7448079>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMSVZIFVOM6DXYYN7E4JU3YCJ6DXAVCNFSM6AAAAAA6ZLX4GSVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TINBYGA3TS>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Mark Rose
***@***.***
925.922.7263

**Reply by jbell-astrobotic** [reply]
Okay, yep I was able to get something like this to work! Just had to install npm on top of eclipse-temurin:11.0.21_9-jre-alpine and then ran via ./bin/yamcsd
